### PR TITLE
Fix issue 18832 — std.typecons.tuple comparison broken for floating point elements

### PR DESCRIPTION
**Pull request by Bolpat**
This code compiles:
```d
import std.typecons : tuple;
auto tup = tuple(double.nan);
assert(tup > tup);
assert(!(tup < tup));
```
The reason is that `Tuple`s `opCmp` assumes comparison on the fields are total.

(Cannot reopen #6498 due to GitHub [issue](https://github.com/isaacs/github/issues/361))

### CODE REVIEW

**Comment by Bolpat** [comment]
I don't know what's wrong with the linter. It says ```*** Error in `/home/circleci/phobos/generated/dscanner-032ac7e3ed5ea7df5e097badcfcd91d3cb8f18da/dsc': free(): invalid pointer: 0x00007ffff7fe4a00 ***``` Maybe it suffices to restart the CI.

**Comment by Bolpat** [comment]
The linter should be revised. It rejects auto functions because of no return statement even if they contain a `mixin` statement.

**Comment by wilzbach** [comment]
> The linter should be revised. It rejects auto functions because of no return statement even if they contain a mixin statement.

Are you referring to DScanner? Please open an issue with a testcase there then? Thanks!

File: std/typecons.d [hunk 1]
```diff
@@ -805,48 +800,69 @@ if (distinctFieldNames!(Specs))
...
-
+        auto opCmp(R)(R rhs) { mixin(opCmpImpl); return 0; }
         /// ditto
-        int opCmp(R)(R rhs) const
-        if (areCompatibleTuples!(typeof(this), R, "<"))
-        {
-            foreach (i, Unused; Types)
-            {
-                if (field[i] != rhs.field[i])
+        auto opCmp(R)(R rhs) const { mixin(opCmpImpl); return 0; }
```
Comment (wilzbach) [inline std/typecons.d]: Should we have a constraint here? (it used to have one)

Comment (wilzbach) [reply std/typecons.d]: and what's the motivation for using a mixin? I don't see any reason

File: std/typecons.d [hunk 2]
```diff
@@ -1190,136 +1206,139 @@ if (distinctFieldNames!(Specs))
...
+                        formatElement(writer, field[i], f);
+                    }
                 }
+                writer(footer);
+            }
+            else
+            {
+                throw new FormatException(
+                    "Expected '%s' or '%(...%)' or '%(...%|...%)' format specifier for type '" ~
+                        Unqual!(typeof(this)).stringof ~ "', not '%" ~ fmt.spec ~ "'.");
```
Comment (wilzbach) [inline std/typecons.d]: AFAICT this isn't related to this PR. Please move it into a separate PR. Thanks!

File: std/typecons.d (hunk 1 above)
Comment (Bolpat) [reply std/typecons.d]: Yes, I forgot the constraint.
Rationale for the mixin: The functions cannot be unified (e.g. with `inout`), so this was an attempt to reduce non-trivial code duplication.

**Comment by Bolpat** [comment]
I'll close this PR in favor of splitting the minor changes off into separate one.
It will be reopened and force pushed after the new PR is merged.

File: std/typecons.d (hunk 2 above)
Comment (Bolpat) [reply std/typecons.d]: I'll do.

**Comment by wilzbach** [comment]
Please squash instead of merging.

**Comment by Bolpat** [comment]
I can do squash and rebase manually, but doesn't Github have the squash and merge feature any more? Some time ago, I learned that manual rebasing wasn't necessary any more.

**Comment by wilzbach** [comment]
Yes it does, but people abused it here, which is we disabled it for dlang-bot

**Comment by Bolpat** [comment]
Technically, there are two distinct changes: Fixing table in the documentation of `toString` and the fix itself. I realized the table is srewed up after the merge of #6502.

**Comment by wilzbach** [comment]
> Technically, there are two distinct changes

Then please separate them in two PRs (or move the doc change fix in another PR). Thanks!

File: std/typecons.d [hunk 3]
```diff
@@ -1328,6 +1349,48 @@ if (distinctFieldNames!(Specs))
...
+        static if (Types.length == 0)
+        @safe unittest
+        {
+            import std.format : format, FormatException;
+            import std.exception : assertThrown;
+
+            // enum tupStr = tuple(1, 1.0).toString; // toString is *impure* for double (issue 17628)
+            // static assert(tupStr == `Tuple!(int, double)(1, 1)`);
+
+            Tuple!(int, double)[3] tupList = [ tuple(1, 1.0), tuple(2, 4.0), tuple(3, 9.0) ];
```
Comment (wilzbach) [inline std/typecons.d]: Wasn't this already part of #6502?

Comment (Bolpat) [reply std/typecons.d]: It should have been. Now, it's in #6504.

File: std/typecons.d [hunk 4]
```diff
@@ -800,48 +800,69 @@ if (distinctFieldNames!(Specs))
...
-                    return field[i] < rhs.field[i] ? -1 : 1;
-                }
-            }
-            return 0;
-        }
-
-        /// ditto
-        int opCmp(R)(R rhs) const
-        if (areCompatibleTuples!(typeof(this), R, "<"))
+        auto opCmp(this L, R)(R rhs)
```
Comment (wilzbach) [inline std/typecons.d]: Misses the template constraint

File: std/typecons.d [hunk 5]
```diff
@@ -1185,6 +1206,7 @@ if (distinctFieldNames!(Specs))
             return h;
         }
 
+
```
Comment (wilzbach) [inline std/typecons.d]: Unrelated

File: std/typecons.d (hunk 4 above)
Comment (Bolpat) [reply std/typecons.d]: Corrected.

File: std/typecons.d (hunk 5 above)
Comment (Bolpat) [reply std/typecons.d]: Removed.

**Comment by Bolpat** [comment]
There is more stuff to be done to fix the ordering.

**Comment by wilzbach** [comment]
Why do you constantly keep closing this PR? That's more noise than keeping it open.

**Comment by Bolpat** [comment]
For the last one: I thought I wouldn't be there some time.
I've added `isPartiallyOrdered` in `std.traits` in #6513. The issue boils down to D not having real support for partially ordered types (*partially* meaning *possibly non-total*, not the mathematical term).
@wilzbach Should I reopen now or when #6513 is merged (if it is not rejected)?
